src/client/Client.cc

```diff
@@ -1729,6 +1729,12 @@ void Client::connect_mds_targets(mds_rank_t mds)
 	mdsmap->is_clientreplay_or_active_or_stopping(rank)) {
       ldout(cct, 10) << "check_mds_sessions opening mds." << mds
 		     << " export target mds." << rank << dendl;
+
+      auto session = _get_or_open_mds_session(rank);
+      if (session->state == MetaSession::STATE_OPENING ||
+          session->state == MetaSession::STATE_OPEN)
+        continue;
```

Comment:
If the session didn’t exist in mds_sessions and its state Still is STATE_OPENINT or STATE_OPEN, which indicate the session has broken. Thus, i think we should send open session request to mds and this also can reset the session state to OPENING.

Reply:
@yunfeiguan Not sure if I understand your comment - do you mean this is buggy?

Reply:
Yes，in my opinion, execute continue will lead the session stay in STATE_POEN of STATE_OPENING forever. What’s more, due to the session didn’t exist in mds_sessions, we can’t use this session and will hang all the time.

Reply:
The `connect_mds_targets()` could be called when a new `mdsmap` comes and there are new `MDSs` change to `ACTIVE` state:

```
diff --git a/src/client/Client.cc b/src/client/Client.cc
index b1cd017a948..b0b28e0528f 100644
--- a/src/client/Client.cc
+++ b/src/client/Client.cc
@@ -2286,6 +2286,7 @@ MetaSessionRef Client::_open_mds_session(mds_rank_t mds)
   m->metadata = metadata;
   m->supported_features = feature_bitset_t(CEPHFS_FEATURES_CLIENT_SUPPORTED);
   m->metric_spec = feature_bitset_t(CEPHFS_METRIC_FEATURES_ALL);
+  session->state = MetaSession::STATE_OPENING;
   session->con->send_message2(std::move(m));
   return session;
 }
diff --git a/src/client/MetaSession.h b/src/client/MetaSession.h
index ad74ae58af2..47d22171a48 100644
--- a/src/client/MetaSession.h
+++ b/src/client/MetaSession.h
@@ -28,14 +28,14 @@ struct MetaSession {
   feature_bitset_t mds_metric_flags;
 
   enum {
-    STATE_NEW, // Unused
+    STATE_NEW,
     STATE_OPENING,
     STATE_OPEN,
     STATE_CLOSING,
     STATE_CLOSED,
     STATE_STALE,
     STATE_REJECTED,
-  } state = STATE_OPENING;
+  } state = STATE_NEW;
 
   enum {
     RECLAIM_NULL,

```

Reply:
> > If the session didn’t exist in mds_sessions and its state Still is STATE_OPENINT or STATE_OPEN, which indicate the session has broken.
> 
> If the session didn't exist in the `mds_sessions`, how could it be possible the state is `STATE_OPEN` ?
> 
> > Thus, i think we should send open session request to mds and this also can reset the session state to OPENING.
> 
> We should set the session state to `STATE_NEW`, which for now it's `STATE_OPENING` as default, when the session is new added to the `mds_sessions` and then set it to `STATE_OPENING` after this.

Sorry, it's okay to set it to `STATE_OPENING`, misread the code.

Reply:
As disscussed in Wechat, the `_get_or_open_mds_session()` will send the open session request. The code above really makes not much sense, but not harm. Let's improve it later with other fixes in future.